README.md
# CITEVizTestData

`CITEVizTestData` is an ExperimentHub package that contains two CITE-Seq datasets from peripheral blood mononuclear cells (PBMC). Each of the datasets are identical, but formatted as a Seurat or SingleCellExperiment object. Each object contains 2500 cells and 228 antibodies, and is a down-sampled version of the data by [Hao et. al. 2021](https://www.sciencedirect.com/science/article/pii/S0092867421005833?via%3Dihub) under the [Creative Commons BY 4.0](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwjOjeLamN76AhWpATQIHQZhDPgQFnoECBEQAQ&url=https%3A%2F%2Fcreativecommons.org%2Flicenses%2Fby%2F4.0%2F&usg=AOvVaw28E0YDTT_NxJYfRVmipfeb) license. Users can have the choice to download the data in either format and export it as an RDS file for use with CITEViz. For more information about CITEViz, please click [here](https://github.com/maxsonBraunLab/CITEViz).

# How to download the data from ExperimentHub

While there are multiple ways to download data from ExperimentHub (more info in the `CITEVizTestData` vignette), below is one such example:

```R
library(ExperimentHub)

# create EH instance and query it
eh <- ExperimentHub()
query_results <- query(eh, "CITEVizTestData")

# get PBMC CITE-Seq datasets in two formats
file_seurat <- query_results[["EH7739"]] # Test data in Seurat format
file_sce <- query_results[["EH7740"]] # Test data in SingleCellExperiment format

```

The next step is to export the data to an RDS file, which can be uploaded to [CITEViz](https://github.com/maxsonBraunLab/CITEViz) via the upload button in the GUI.

```R
# export information to external file
saveRDS(file_seurat, "CITEVizTestData_Seurat.rds")
saveRDS(file_sce, "CITEVizTestData_SingleCellExperiment.rds")

```


inst/NEWS
changes in version 0.99.0 (2022-10-13)
+ initial submission to Bioconductor

R/CITEVizTestData-pkg.R
#' CITEVizTestData: Example CITE-Seq datasets for CITEViz
#'
#' \code{CITEVizTestData} is an ExperimentHub package containing 
#' data for use with CITEViz. CITEViz is an RShiny app that allows users to 
#' classify cell types in CITE-Seq data via cellular surface protein levels. 
#' Similar to flow cytometry, users can physically draw gates and filter on a
#' mixture of cells, and the selected cells are immediately highlighted in
#' dimensional reduction space (e.g. UMAP, tSNE, PCA). This workflow provides
#' rapid feedback to the users to help classify cell types. To read more about
#' CITEViz, click (here. \url{https://github.com/maxsonBraunLab/CITEViz})
#' 
#' The input of CITEViz is a pre-processed Seurat/SingleCelExperiment object
#' with assays containing normalized RNA/SCT (single-cell transform)
#' and ADT (antibody-derived tag) count data, as well as data for at least
#' one dimensional reduction (e.g. UMAP, PCA, etc.). The CITEVizTestData 
#' package allows users to quickly download an example dataset for the CITEViz
#' tutorial.
#' 
#' The available datasets are CITE-Seq assays of human peripheral bone and
#' mononuclear cells with 2500 cells and 228 antibodies from
#' (Hao et. al. 2021, \url{https://linkinghub.elsevier.com/retrieve/pii/S0092-8674(21)00583-3})
#' made available through the (Creative Commons BY 4.0, \url{https://creativecommons.org/licenses/by/4.0/})
#' license.
#' @author Garth Kong \email{kongga2017@gmail.com}
#' @name CITEVizTestData-pkg
#' @docType package
NULL

inst/scripts/make-data.R
library(Seurat)
library(SeuratDisk)
library(SingleCellExperiment)

# define the data
url="https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat"
filename="pbmc_multimodal.h5seurat"

# download the file
utils::download.file(url, destfile = filename, method = "curl")

# import and down-sample
pbmc = SeuratDisk::LoadH5Seurat(filename) # this is only recommended for computers with > 16GB RAM

# get cell barcodes
barcodes = rownames(pbmc@meta.data)

# random sample cell barcodes
set.seed(1)
sampled_barcodes = sample(barcodes, 2500)

# subset pbmc data
small_pbmc = subset(pbmc, cells = sampled_barcodes)

# export as Seurat object
saveRDS(small_pbmc, "pbmc_2500.rds")

# export as SCE object
small_pbmc_sce = as.SingleCellExperiment(pbmc)
saveRDS(small_pbmc, "pbmc_2500_sce.rds")

inst/scripts/make-metadata.R
url = "https://atlas.fredhutch.org/data/nygc/multimodal/pbmc_multimodal.h5seurat"
md5sum = "82421ea2c16fe30fba15559efbcf9094"
species = "Homo sapiens"

d <- data.frame(
	Title = c("CITEVizTestDataSeurat", "CITEVizTestDataSCE"),
	Description = c("PBMC CITE-Seq data in Seurat format with 2500 cells", "PBMC CITE-Seq data in SingleCellExperiment format with 2500 cells"),
	BiocVersion = rep("3.16", 2),
	Genome = rep("GRCh38", 2),
	SourceType = rep("HDF5", 2),
	SourceUrl = rep(url, 2),
	SourceVersion = rep(md5sum, 2),
	Species = rep(species, 2),
	TaxonomyId = rep(9606, 2),
	Coordinate_1_based = rep(TRUE, 2),
	DataProvider = rep("Satija Lab", 2),
	Maintainer = rep("Garth Kong kongga2017@gmail.com", 2),
	RDataClass = c("Seurat", "SingleCellExperiment"),
	DispatchClass = rep("Rds", 2),
	stringsAsFactors = FALSE,
	RDataPath = c("CITEVizTestData/pbmc_2500.rds", "CITEVizTestData/pbmc_2500_sce.rds")
)

write.csv(d, "inst/extdata/metadata.csv", row.names = FALSE)

R/CITEVizTestData.R
#' @title Download a PBMC CITE-Seq data for CITEViz in 
#' Seurat/SingleCellExperiment format using ExperimentHub
#' 
#' @description This function helps download a CITE-Seq dataset from 
#' Hao. et. al. 2021 for use with CITEViz. The dataset contains 
#' 2500 cells and 228 antibodies.
#'
#' @param format Either "Seurat" or "SingleCellExperiment"
#' @return A Seurat [default] or SingleCellExperiment object.
#' @examples \donttest{
#' library(CITEVizTestData)
#' file_seurat = CITEVizTestData(format = "Seurat")
#' file_sce = CITEVizTestData(format = "SingleCellExperiment")
#' }
#' 
#' @importFrom ExperimentHub ExperimentHub
#' @importFrom AnnotationHub query
#' 
#' @export
CITEVizTestData <- function(format = "Seurat") {
  
  all_datasets <- c("CITEVizTestDataSeurat",
                    "CITEVizTestDataSingleCellExperiment")
  all_formats <- c("Seurat", "SingleCellExperiment")
  
  if (format %in% all_formats) {
  
    eh = ExperimentHub::ExperimentHub()
    pkgname <- "CITEVizTestData"
    q <- AnnotationHub::query(eh, pkgname)
    
    if (format == "Seurat") {
      # I know I am not supposed to use the index, 
      # but I currently don't have an EHID right now.
      out <- q[["EH7739"]]
    }
    
    if (format == "SingleCellExperiment")
      out <- q[["EH7740"]]
  
  return(out)
    
  } else {
    
    message("Input format argument could not be found.")
    stop("Available formats: ['Seurat', 'SingleCellExperiment']")
  
  }
}

